src/types/simple.rs

```diff
@@ -47,11 +48,55 @@ impl Display for SimpleType {
     }
 }
 
+/// Categorizes types into three classes according to basic operations supported.
+#[derive(
+    Copy, Clone, PartialEq, Eq, Hash, Debug, derive_more::Display, Serialize_repr, Deserialize_repr,
+)]
+#[repr(u8)]
+pub enum TypeClass {
+    /// Any value, including linear and quantum values;
```

Comment:
I'm nervous of calling this `Any` because it will not be the least-upper-bound / union once we have the fourth TypeClass corresponding to TypeArg/TypeParam (that is: Any is *not* a superset of TypeParam). We dislike Linear ;-) but maybe "AnyValue" or "Value" ??